Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/DeleteDriver.hs

```diff
@@ -0,0 +1,84 @@
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module SharedLogic.DeleteDriver where
+
+import qualified Domain.Types.Merchant as DM
+import qualified Domain.Types.Person as DP
+import Environment
+import Kernel.Prelude
+import qualified Kernel.Storage.Esqueleto as Esq
+import qualified Kernel.Storage.Hedis as Redis
+import Kernel.Types.APISuccess (APISuccess (Success))
+import Kernel.Types.Id
+import Kernel.Utils.Common
+import SharedLogic.Merchant (findMerchantByShortId)
+import Storage.CachedQueries.CacheConfig
+import qualified Storage.CachedQueries.DriverInformation as CQDriverInfo
+import qualified Storage.Queries.Driver.DriverFlowStatus as QDriverFlowStatus
+import qualified Storage.Queries.DriverInformation as QDriverInfo
+import qualified Storage.Queries.DriverLocation as QDriverLocation
+import qualified Storage.Queries.DriverOnboarding.DriverLicense as QDriverLicense
+import qualified Storage.Queries.DriverOnboarding.DriverRCAssociation as QRCAssociation
+import qualified Storage.Queries.DriverOnboarding.IdfyVerification as QIV
+import qualified Storage.Queries.DriverOnboarding.Image as QImage
+import qualified Storage.Queries.DriverQuote as QDriverQuote
+import qualified Storage.Queries.DriverStats as QDriverStats
+import qualified Storage.Queries.Message.MessageReport as QMessage
+import qualified Storage.Queries.Person as QPerson
+import qualified Storage.Queries.RegistrationToken as QR
+import qualified Storage.Queries.Ride as QRide
+import qualified Storage.Queries.SearchRequestForDriver as QSearchReqForDriver
+import qualified Storage.Queries.Vehicle as QVehicle
+import qualified Tools.Auth as Auth
+import Tools.Error
+
+deleteDriver :: ShortId DM.Merchant -> Id DP.Person -> Flow APISuccess
+deleteDriver merchantShortId reqDriverId = do
+  merchant <- findMerchantByShortId merchantShortId
+  driver <-
+    QPerson.findById reqDriverId
+      >>= fromMaybeM (PersonDoesNotExist reqDriverId.getId)
+
+  -- merchant access checking
+  merchantDeleteCheck <- merchantAccessCheck merchant driver
+  when merchantDeleteCheck $ throwError $ InvalidRequest "Driver can't be deleted"
+
+  -- this function uses tokens from db, so should be called before transaction
+  Auth.clearDriverSession reqDriverId
+  Esq.runTransaction $ do
+    QIV.deleteByPersonId reqDriverId
+    QImage.deleteByPersonId reqDriverId
+    QDriverLicense.deleteByDriverId reqDriverId
+    QRCAssociation.deleteByDriverId reqDriverId
+    QDriverQuote.deleteByDriverId reqDriverId
+    QSearchReqForDriver.deleteByDriverId reqDriverId
+    QDriverStats.deleteById (cast reqDriverId)
+    QDriverLocation.deleteById reqDriverId
+    QR.deleteByPersonId reqDriverId
+    QVehicle.deleteById reqDriverId
+    QDriverInfo.deleteById (cast reqDriverId)
+    QDriverFlowStatus.deleteById reqDriverId
+    QMessage.deleteByPersonId reqDriverId
+    QPerson.deleteById reqDriverId
```

Comment:
Why don't we have deletetion from Ride table among these queries? Don't we have Ride -> Person ref? If usage of this func expected to be only on drivers, which don't have any rides, then we should indicate it somehow in error or something like that.
Personally i think it's not a good idea to let anyone except Dashboard user to delete driver accounts, but that is up to what you want... We should add a feature to disable drivers without removing the entry from db entirely (mb with removing personal infos like a name and phone num)

Reply:
This function is used for the reason that during registration if the driver has completed login using one number and he tries to login using another number but while uploading DL/RC he uploads same as previous one so we don't allow him but the data is saved as another driver's data with mobile number (incomplete registration) so if he enters his this number as alternate number previously since the number is present in Db we were not allowing him but now we are using this function to delete the driver who passes the checks for deletion and allow driver to enter alternate number

Reply:
@DonDokidon Correct. We should not delete any driver other than dashboard. I was also thinking the same. 
But this is very tricky case. We have introduced alternate number flow recently where a driver can have an alternate number considering that alternate number is not registered with us. But in case, driver mistakenly logged in using his alternate number then he will not be able to add this number as it is already registered. 

To handle this case, driver should call our operations team and get this extra entry deleted. But as this will be very operational heavy and we don't have enough number of operations staff, so we're automating this process by deleting it after OTP verification.